File: nodejs/xmlApiUtils.js
/**
 * Global Payments XML API Utilities
 *
 * This module provides utility functions for XML API integration including:
 * - SHA-1 hash generation for request authentication
 * - Timestamp formatting
 * - XML request building
 * - XML response parsing
 */

import crypto from 'crypto';
import { parseString, Builder } from 'xml2js';

/**
 * Generate timestamp in XML API format: YYYYMMDDHHMMSS
 * @returns {string} Formatted timestamp
 */
export function generateTimestamp() {
    const now = new Date();
    const year = now.getFullYear();
    const month = String(now.getMonth() + 1).padStart(2, '0');
    const day = String(now.getDate()).padStart(2, '0');
    const hours = String(now.getHours()).padStart(2, '0');
    const minutes = String(now.getMinutes()).padStart(2, '0');
    const seconds = String(now.getSeconds()).padStart(2, '0');

    return `${year}${month}${day}${hours}${minutes}${seconds}`;
}

/**
 * Generate unique order ID
 * @param {string} prefix - Optional prefix for order ID
 * @returns {string} Unique order ID
 */
export function generateOrderId(prefix = 'ORD') {
    const timestamp = Date.now();
    const random = Math.floor(Math.random() * 1000);
    return `${prefix}_${timestamp}_${random}`;
}

/**
 * Generate SHA-1 hash for XML API authentication
 * XML API uses a two-step hashing process:
 * 1. Hash the concatenated request values
 * 2. Hash the result with the shared secret appended
 *
 * @param {string} dataString - Concatenated request values (e.g., "timestamp.merchantid.orderid.amount.currency.cardnumber")
 * @param {string} sharedSecret - Merchant shared secret
 * @returns {string} SHA-1 hash for authentication
 */
export function generateSHA1Hash(dataString, sharedSecret) {
    // Step 1: Hash the data string
    const firstHash = crypto
        .createHash('sha1')
        .update(dataString)
        .digest('hex');

    // Step 2: Concatenate first hash with shared secret and hash again
    const secondHash = crypto
        .createHash('sha1')
        .update(`${firstHash}.${sharedSecret}`)
        .digest('hex');

    return secondHash;
}

/**
 * Generate authentication hash for payment request
 * Hash blueprint: timestamp.merchantid.orderid.amount.currency.cardnumber
 *
 * @param {Object} params - Request parameters
 * @param {string} params.timestamp - Request timestamp
 * @param {string} params.merchantId - Merchant ID
 * @param {string} params.orderId - Order ID
 * @param {string} params.amount - Amount in smallest currency unit (e.g., cents)
 * @param {string} params.currency - Currency code (e.g., 'USD')
 * @param {string} params.cardNumber - Card number (optional, empty string for token-based)
 * @param {string} sharedSecret - Merchant shared secret
 * @returns {string} Authentication hash
 */
export function generatePaymentHash(params, sharedSecret) {
    const { timestamp, merchantId, orderId, amount, currency, cardNumber = '' } = params;
    const dataString = `${timestamp}.${merchantId}.${orderId}.${amount}.${currency}.${cardNumber}`;
    return generateSHA1Hash(dataString, sharedSecret);
}

/**
 * Generate authentication hash for stored card payment (receipt-in)
 * Hash blueprint: timestamp.merchantid.orderid.amount.currency.payerref
 *
 * @param {Object} params - Request parameters
 * @param {string} params.timestamp - Request timestamp
 * @param {string} params.merchantId - Merchant ID
 * @param {string} params.orderId - Order ID
 * @param {string} params.amount - Amount in smallest currency unit
 * @param {string} params.currency - Currency code
 * @param {string} params.payerRef - Customer/payer reference
 * @param {string} sharedSecret - Merchant shared secret
 * @returns {string} Authentication hash
 */
export function generateStoredCardPaymentHash(params, sharedSecret) {
    const { timestamp, merchantId, orderId, amount, currency, payerRef } = params;
    const dataString = `${timestamp}.${merchantId}.${orderId}.${amount}.${currency}.${payerRef}`;
    return generateSHA1Hash(dataString, sharedSecret);
}

/**
 * Generate authentication hash for card storage operations
 * Hash blueprint: timestamp.merchantid.orderid.amount.currency.payerref
 *
 * @param {Object} params - Request parameters
 * @param {string} params.timestamp - Request timestamp
 * @param {string} params.merchantId - Merchant ID
 * @param {string} params.orderId - Order ID
 * @param {string} params.amount - Amount (use empty string for storage without payment)
 * @param {string} params.currency - Currency (use empty string for storage without payment)
 * @param {string} params.payerRef - Customer/payer reference
 * @param {string} sharedSecret - Merchant shared secret
 * @returns {string} Authentication hash
 */
export function generateCardStorageHash(params, sharedSecret) {
    const { timestamp, merchantId, orderId, amount = '', currency = '', payerRef } = params;
    const dataString = `${timestamp}.${merchantId}.${orderId}.${amount}.${currency}.${payerRef}`;
    return generateSHA1Hash(dataString, sharedSecret);
}

/**
 * Generate authentication hash for Payment Scheduler operations
 * Hash blueprint: timestamp.merchantid.scheduleref.amount.currency.payerref.schedule
 *
 * @param {Object} params - Request parameters
 * @param {string} params.timestamp - Request timestamp
 * @param {string} params.merchantId - Merchant ID
 * @param {string} params.scheduleRef - Schedule reference
 * @param {string} params.amount - Amount in smallest currency unit
 * @param {string} params.currency - Currency code
 * @param {string} params.payerRef - Customer/payer reference
 * @param {string} params.schedule - Schedule frequency (e.g., 'monthly', 'weekly')
 * @param {string} sharedSecret - Merchant shared secret
 * @returns {string} Authentication hash
 */
export function generateScheduleHash(params, sharedSecret) {
    const { timestamp, merchantId, scheduleRef, amount, currency, payerRef, schedule } = params;
    const dataString = `${timestamp}.${merchantId}.${scheduleRef}.${amount}.${currency}.${payerRef}.${schedule}`;
    return generateSHA1Hash(dataString, sharedSecret);
}

/**
 * Verify response hash from XML API
 * Response hash blueprint: timestamp.merchantid.orderid.result.message.pasref.authcode
 *
 * @param {Object} response - Response object from XML API
 * @param {string} sharedSecret - Merchant shared secret
 * @returns {boolean} True if hash is valid
 */
export function verifyResponseHash(response, sharedSecret) {
    const { timestamp, merchantid, orderid, result, message, pasref = '', authcode = '' } = response;
    const dataString = `${timestamp}.${merchantid}.${orderid}.${result}.${message}.${pasref}.${authcode}`;
    const calculatedHash = generateSHA1Hash(dataString, sharedSecret);

    return calculatedHash === response.sha1hash;
}

/**
 * Convert amount to smallest currency unit (cents)
 * XML API requires amounts without decimal points
 *
 * @param {number} amount - Amount in dollars (e.g., 29.99)
 * @returns {string} Amount in cents (e.g., '2999')
 */
export function convertToCents(amount) {
    return String(Math.round(amount * 100));
}

/**
 * Convert amount from smallest currency unit to dollars
 *
 * @param {string|number} cents - Amount in cents
 * @returns {number} Amount in dollars
 */
export function convertFromCents(cents) {
    return Number(cents) / 100;
}

/**
 * Sanitize postal code for XML API
 * Allowed characters: alphanumeric, hyphen, space
 *
 * @param {string} postalCode - Postal code to sanitize
 * @returns {string} Sanitized postal code
 */
export function sanitizePostalCode(postalCode) {
    if (!postalCode) return '';
    return postalCode.replace(/[^a-zA-Z0-9- ]/g, '').slice(0, 10);
}

/**
 * Sanitize alphanumeric fields for XML API
 * Allowed characters: alphanumeric, /.-_', and space
 *
 * @param {string} value - Value to sanitize
 * @param {number} maxLength - Maximum length
 * @returns {string} Sanitized value
 */
export function sanitizeAlphanumeric(value, maxLength = 255) {
    if (!value) return '';
    return value.replace(/[^a-zA-Z0-9\/.\-_', ]/g, '').slice(0, maxLength);
}

/**
 * Parse XML response from XML API
 *
 * @param {string} xmlString - XML response string
 * @returns {Promise<Object>} Parsed response object
 */
export async function parseXMLResponse(xmlString) {
    return new Promise((resolve, reject) => {
        parseString(xmlString, { explicitArray: false, mergeAttrs: true }, (err, result) => {
            if (err) {
                reject(new Error(`XML parsing error: ${err.message}`));
            } else {
                resolve(result.response);
            }
        });
    });
}

/**
 * Build XML request for XML API
 *
 * @param {Object} requestData - Request data object
 * @returns {string} XML request string
 */
export function buildXMLRequest(requestData) {
    const builder = new Builder({
        xmldec: { version: '1.0', encoding: 'UTF-8' },
        renderOpts: { pretty: false }
    });

    return builder.buildObject({ request: requestData });
}

/**
 * Map frequency string to XML API schedule format
 *
 * @param {string} frequency - Frequency (weekly, bi-weekly, monthly, quarterly, yearly)
 * @returns {string} XML API schedule macro
 */
export function mapFrequencyToSchedule(frequency) {
    const frequencyMap = {
        'weekly': 'weekly',
        'bi-weekly': '? * 6',  // Every Saturday (cron format)
        'biweekly': '? * 6',
        'monthly': 'monthly',
        'quarterly': 'quarterly',
        'yearly': 'yearly',
        'annually': 'yearly'
    };

    return frequencyMap[frequency.toLowerCase()] || 'monthly';
}

/**
 * Format start date for XML API (YYYYMMDD)
 *
 * @param {string|Date} date - Start date (ISO string or Date object)
 * @returns {string} Formatted date (YYYYMMDD)
 */
export function formatStartDate(date) {
    const d = typeof date === 'string' ? new Date(date) : date;
    const year = d.getFullYear();
    const month = String(d.getMonth() + 1).padStart(2, '0');
    const day = String(d.getDate()).padStart(2, '0');

    return `${year}${month}${day}`;
}

/**
 * Get XML API endpoint URL based on environment
 *
 * @param {string} environment - 'sandbox' or 'production'
 * @returns {string} XML API endpoint URL
 */
export function getXMLAPIEndpoint(environment = 'sandbox') {
    return environment === 'production'
        ? 'https://api.realexpayments.com/epage-remote.cgi'
        : 'https://api.sandbox.realexpayments.com/epage-remote.cgi';
}

/**
 * Parse XML API error response
 *
 * @param {Object} response - Parsed XML API response
 * @returns {Object} Error information
 */
export function parseErrorResponse(response) {
    return {
        result: response.result,
        message: response.message || 'Unknown error',
        timestamp: response.timestamp,
        orderId: response.orderid
    };
}

File: java/src/main/java/com/globalpayments/xmlapi/XmlApiUtils.java
package com.globalpayments.xmlapi;

import java.io.StringReader;
import java.io.StringWriter;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;

/**
 * Global Payments XML API Utilities
 *
 * Provides utility functions for XML API integration including:
 * - SHA-1 hash generation for request authentication
 * - Timestamp formatting
 * - XML request building
 * - XML response parsing
 * - Country code conversion
 */
public class XmlApiUtils {

    private static final DateTimeFormatter TIMESTAMP_FORMATTER = DateTimeFormatter.ofPattern("yyyyMMddHHmmss");
    private static final Random RANDOM = new Random();

    /**
     * Generate timestamp in XML API format: YYYYMMDDHHMMSS
     */
    public static String generateTimestamp() {
        return LocalDateTime.now().format(TIMESTAMP_FORMATTER);
    }

    /**
     * Generate unique order ID
     */
    public static String generateOrderId(String prefix) {
        long timestamp = System.currentTimeMillis();
        int randomNum = RANDOM.nextInt(1000);
        return String.format("%s_%d_%d", prefix, timestamp, randomNum);
    }

    public static String generateOrderId() {
        return generateOrderId("ORD");
    }

    /**
     * Generate SHA-1 hash for XML API authentication
     *
     * XML API uses a two-step hashing process:
     * 1. Hash the concatenated request values
     * 2. Hash the result with the shared secret appended
     */
    public static String generateSha1Hash(String dataString, String sharedSecret) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-1");

            // First hash
            byte[] firstHashBytes = md.digest(dataString.getBytes(StandardCharsets.UTF_8));
            String firstHash = bytesToHex(firstHashBytes);

            // Second hash with shared secret
            String secondString = firstHash + "." + sharedSecret;
            byte[] secondHashBytes = md.digest(secondString.getBytes(StandardCharsets.UTF_8));
            String finalHash = bytesToHex(secondHashBytes);

            return finalHash;
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-1 algorithm not available", e);
        }
    }

    /**
     * Convert byte array to hex string
     */
    private static String bytesToHex(byte[] bytes) {
        StringBuilder hexString = new StringBuilder();
        for (byte b : bytes) {
            String hex = Integer.toHexString(0xff & b);
            if (hex.length() == 1) {
                hexString.append('0');
            }
            hexString.append(hex);
        }
        return hexString.toString();
    }

    /**
     * Generate hash for payer-new request
     */
    public static String generatePayerNewHash(Map<String, String> params, String sharedSecret) {
        String dataString = String.format("%s.%s.%s...%s",
            params.get("timestamp"),
            params.get("merchantId"),
            params.get("orderId"),
            params.get("payerRef")
        );
        return generateSha1Hash(dataString, sharedSecret);
    }

    /**
     * Generate hash for card-new request
     */
    public static String generateCardNewHash(Map<String, String> params, String sharedSecret) {
        String dataString = String.format("%s.%s.%s...%s.%s.%s",
            params.get("timestamp"),
            params.get("merchantId"),
            params.get("orderId"),
            params.get("payerRef"),
            params.get("chname"),
            params.get("cardNumber")
        );
        return generateSha1Hash(dataString, sharedSecret);
    }

    /**
     * Generate hash for receipt-in (stored card payment) request
     */
    public static String generateStoredCardPaymentHash(Map<String, String> params, String sharedSecret) {
        String dataString = String.format("%s.%s.%s.%s.%s.%s",
            params.get("timestamp"),
            params.get("merchantId"),
            params.get("orderId"),
            params.get("amount"),
            params.get("currency"),
            params.get("payerRef")
        );
        return generateSha1Hash(dataString, sharedSecret);
    }

    /**
     * Generate hash for schedule-new request
     */
    public static String generateScheduleHash(Map<String, String> params, String sharedSecret) {
        String dataString = String.format("%s.%s.%s.%s.%s.%s.%s",
            params.get("timestamp"),
            params.get("merchantId"),
            params.get("scheduleRef"),
            params.get("amount"),
            params.get("currency"),
            params.get("payerRef"),
            params.get("schedule")
        );
        return generateSha1Hash(dataString, sharedSecret);
    }

    /**
     * Convert dollar amount to cents
     */
    public static int convertToCents(double amount) {
        return (int) Math.round(amount * 100);
    }

    /**
     * Convert cents to dollars
     */
    public static double convertFromCents(int cents) {
        return cents / 100.0;
    }

    /**
     * Sanitize postal code
     */
    public static String sanitizePostalCode(String postalCode) {
        if (postalCode == null || postalCode.isEmpty()) {
            return "";
        }
        return postalCode.replace(" ", "").replace("-", "").substring(0, Math.min(postalCode.length(), 16));
    }

    /**
     * Sanitize alphanumeric string
     */
    public static String sanitizeAlphanumeric(String value, int maxLength) {
        if (value == null || value.isEmpty()) {
            return "";
        }
        // Remove any potentially problematic characters
        String sanitized = value.replaceAll("[^a-zA-Z0-9 .,@_-]", "");
        return sanitized.substring(0, Math.min(sanitized.length(), maxLength));
    }

    public static String sanitizeAlphanumeric(String value) {
        return sanitizeAlphanumeric(value, 255);
    }

    /**
     * Map frequency to schedule format
     */
    public static String mapFrequencyToSchedule(String frequency) {
        Map<String, String> frequencyMap = new HashMap<>();
        frequencyMap.put("weekly", "weekly");
        frequencyMap.put("bi-weekly", "biweekly");
        frequencyMap.put("biweekly", "biweekly");
        frequencyMap.put("monthly", "monthly");
        frequencyMap.put("quarterly", "quarterly");
        frequencyMap.put("yearly", "yearly");
        frequencyMap.put("annually", "yearly");

        return frequencyMap.getOrDefault(frequency.toLowerCase(), "monthly");
    }

    /**
     * Get XML API endpoint based on environment
     */
    public static String getXmlApiEndpoint(String environment) {
        if ("production".equals(environment)) {
            return "https://api.realexpayments.com/epage-remote.cgi";
        }
        return "https://api.sandbox.realexpayments.com/epage-remote.cgi";
    }

    public static String getXmlApiEndpoint() {
        return getXmlApiEndpoint("sandbox");
    }

    /**
     * Convert ISO 3166-1 numeric country code to alpha-2 code
     * HPP uses numeric codes (e.g., "840"), but XML API uses alpha-2 codes (e.g., "US")
     */
    public static String convertCountryCodeToAlpha2(String numericCode) {
        Map<String, String> countryCodeMap = new HashMap<>();
        countryCodeMap.put("036", "AU"); // Australia
        countryCodeMap.put("124", "CA"); // Canada
        countryCodeMap.put("276", "DE"); // Germany
        countryCodeMap.put("372", "IE"); // Ireland
        countryCodeMap.put("826", "GB"); // United Kingdom
        countryCodeMap.put("840", "US"); // United States
        countryCodeMap.put("250", "FR"); // France
        countryCodeMap.put("380", "IT"); // Italy
        countryCodeMap.put("724", "ES"); // Spain
        countryCodeMap.put("528", "NL"); // Netherlands
        countryCodeMap.put("056", "BE"); // Belgium
        countryCodeMap.put("208", "DK"); // Denmark
        countryCodeMap.put("246", "FI"); // Finland
        countryCodeMap.put("578", "NO"); // Norway
        countryCodeMap.put("752", "SE"); // Sweden
        countryCodeMap.put("756", "CH"); // Switzerland
        countryCodeMap.put("040", "AT"); // Austria
        countryCodeMap.put("620", "PT"); // Portugal
        countryCodeMap.put("616", "PL"); // Poland
        countryCodeMap.put("203", "CZ"); // Czech Republic

        // If already alpha-2 (2 characters), return as-is
        if (numericCode != null && numericCode.length() == 2) {
            return numericCode.toUpperCase();
        }

        // Convert numeric to alpha-2
        return countryCodeMap.getOrDefault(numericCode, "US");
    }

    /**
     * Build XML request string from Map
     */
    public static String buildXmlRequest(String requestType, Map<String, Object> requestData) {
        try {
            DocumentBuilderFactory docFactory = DocumentBuilderFactory.newInstance();
            DocumentBuilder docBuilder = docFactory.newDocumentBuilder();
            Document doc = docBuilder.newDocument();

            Element rootElement = doc.createElement("request");
            rootElement.setAttribute("type", requestType);
            rootElement.setAttribute("timestamp", (String) requestData.get("timestamp"));
            doc.appendChild(rootElement);

            // Add elements from requestData
            addElementsToXml(doc, rootElement, requestData);

            // Convert Document to String
            TransformerFactory transformerFactory = TransformerFactory.newInstance();
            Transformer transformer = transformerFactory.newTransformer();
            transformer.setOutputProperty(OutputKeys.INDENT, "no");
            transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");

            StringWriter writer = new StringWriter();
            transformer.transform(new DOMSource(doc), new StreamResult(writer));

            return writer.toString();
        } catch (Exception e) {
            throw new RuntimeException("Failed to build XML request", e);
        }
    }

    @SuppressWarnings("unchecked")
    private static void addElementsToXml(Document doc, Element parent, Map<String, Object> data) {
        for (Map.Entry<String, Object> entry : data.entrySet()) {
            String key = entry.getKey();
            Object value = entry.getValue();

            // Skip special keys
            if ("timestamp".equals(key)) continue;

            if (value instanceof Map) {
                Element element = doc.createElement(key);
                parent.appendChild(element);
                addElementsToXml(doc, element, (Map<String, Object>) value);
            } else if (value instanceof String) {
                Element element = doc.createElement(key);
                element.setTextContent((String) value);
                parent.appendChild(element);
            }
        }
    }

    /**
     * Parse XML response string
     */
    public static Map<String, Object> parseXmlResponse(String xmlString) {
        try {
            DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
            DocumentBuilder builder = factory.newDocumentBuilder();
            Document doc = builder.parse(new InputSource(new StringReader(xmlString)));

            Element root = doc.getDocumentElement();
            return elementToMap(root);
        } catch (Exception e) {
            throw new RuntimeException("Failed to parse XML response", e);
        }
    }

    private static Map<String, Object> elementToMap(Element element) {
        Map<String, Object> result = new HashMap<>();

        // Add attributes
        if (element.hasAttributes()) {
            for (int i = 0; i < element.getAttributes().getLength(); i++) {
                Node attr = element.getAttributes().item(i);
                result.put(attr.getNodeName(), attr.getNodeValue());
            }
        }

        // Add child elements
        NodeList children = element.getChildNodes();
        for (int i = 0; i < children.getLength(); i++) {
            Node child = children.item(i);
            if (child.getNodeType() == Node.ELEMENT_NODE) {
                Element childElement = (Element) child;
                String key = childElement.getTagName();

                // If child has children, recurse
                if (childElement.hasChildNodes() && childElement.getElementsByTagName("*").getLength() > 0) {
                    result.put(key, elementToMap(childElement));
                } else {
                    result.put(key, childElement.getTextContent());
                }
            }
        }

        return result;
    }

    /**
     * Normalize card expiry format
     */
    public static Map<String, String> normalizeCardExpiry(Map<String, String> cardDetails) {
        if (cardDetails == null) {
            return cardDetails;
        }

        Map<String, String> normalized = new HashMap<>(cardDetails);

        // Extract month
        String month = cardDetails.getOrDefault("expMonth",
                      cardDetails.getOrDefault("expmonth",
                      cardDetails.getOrDefault("exp_month",
                      cardDetails.getOrDefault("EXPMONTH", ""))));

        // Extract year
        String year = cardDetails.getOrDefault("expYear",
                     cardDetails.getOrDefault("expyear",
                     cardDetails.getOrDefault("exp_year",
                     cardDetails.getOrDefault("EXPYEAR", ""))));

        // Normalize year to 2 digits
        if (!year.isEmpty() && year.length() == 4) {
            year = year.substring(2);
        }

        if (!month.isEmpty()) {
            normalized.put("expmonth", String.format("%02d", Integer.parseInt(month)));
            normalized.remove("expMonth");
            normalized.remove("exp_month");
            normalized.remove("EXPMONTH");
        }

        if (!year.isEmpty()) {
            normalized.put("expyear", String.format("%02d", Integer.parseInt(year)));
            normalized.remove("expYear");
            normalized.remove("exp_year");
            normalized.remove("EXPYEAR");
        }

        return normalized;
    }
}
